File: src/main/java/kimspring/splearn/adapter/webapi/MemberApi.java
package kimspring.splearn.adapter.webapi;

import jakarta.validation.Valid;
import kimspring.splearn.adapter.webapi.dto.MemberRegisterResponse;
import kimspring.splearn.application.member.provided.MemberRegister;
import kimspring.splearn.domain.member.Member;
import kimspring.splearn.domain.member.MemberRegisterRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
public class MemberApi {
    private final MemberRegister memberRegister;

    @PostMapping("/api/members")
    public MemberRegisterResponse register(@RequestBody @Valid MemberRegisterRequest request) {
        Member member = memberRegister.register(request);

        return MemberRegisterResponse.of(member);
    }
}

File: src/main/kotlin/kimspring/splearn/adapter/webapi/MemberApi.kt
package kimspring.splearn.adapter.webapi

import kimspring.splearn.adapter.webapi.dto.MemberRegisterResponse
import kimspring.splearn.adapter.webapi.dto.MemberResponse
import kimspring.splearn.application.member.command.RegisterMemberCommand
import kimspring.splearn.application.member.command.UpdateMemberInfoCommand
import kimspring.splearn.application.member.usecase.MemberLifecycle
import kimspring.splearn.application.member.usecase.MemberModifier
import kimspring.splearn.application.member.usecase.MemberRegister
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RestController
import java.net.URI

@RestController
class MemberApi(
    private val memberRegister: MemberRegister,
    private val memberLifecycle: MemberLifecycle,
    private val memberModifier: MemberModifier,
) : MemberApiSpec {
    override fun register(request: RegisterMemberCommand): ResponseEntity<MemberRegisterResponse> {
        val member = memberRegister.register(request)
        val location = URI.create("/api/members/${requireNotNull(member.id)}")
        return ResponseEntity.created(location).body(MemberRegisterResponse.of(member))
    }

    override fun activate(memberId: Long): MemberResponse = MemberResponse.of(memberLifecycle.activate(memberId))

    override fun deactivate(memberId: Long): MemberResponse = MemberResponse.of(memberLifecycle.deactivate(memberId))

    override fun updateInfo(
        memberId: Long,
        request: UpdateMemberInfoCommand,
    ): MemberResponse = MemberResponse.of(memberModifier.updateInfo(memberId, request))
}
